scipy/linalg/_decomp_lu_cython.pyx

```diff
@@ -0,0 +1,160 @@
+# cython: language_level=3
+
+import cython
+from cpython.mem cimport PyMem_Malloc, PyMem_Free
+from libc.string cimport memcpy
+
+from scipy.linalg.cython_lapack cimport sgetrf, dgetrf, cgetrf, zgetrf
+from scipy.linalg._cythonized_array_utils cimport lapack_t, swap_c_and_f_layout
+
+import numpy as np
+cimport numpy as cnp
+cnp.import_array()
+
+
+ctypedef double complex doubleComplex
+ctypedef float complex floatComplex
+
+
+@cython.nonecheck(False)
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.initializedcheck(False)
+cdef void lu_decompose(cnp.ndarray[lapack_t, ndim=2] a,
+                       cnp.ndarray[lapack_t, ndim=2] lu,
+                       int[::1] perm,
+                       bint permute_l):
```

Comment:
```suggestion
                       bint permute_l) noexcept:
```

Reply:
Just to make sure I understand correctly there is a possibility for this to raise exception below. Is this still needed here?

Reply:
> there is a possibility for this to raise exception below. Is this still needed here?

If there is possibility to have exception raised **and** you want to have correct code for both Cython 0.29.X and 3 you need to explicitly declare exception propagation using `except` keyword.

If you are writing code only for Cython 3 you can avoid specifying anything. I think it is worth to read Cython documentation about it: https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#error-return-values It was pretty improved during recent days.

> **Warning**: using `void` return value can be dangerous when you forgot to provide `noexcept` mainly when `nogil` is also used - see following paragraf from cython documentation:
>
> There’s a known performance pitfall when combining nogil and except * @cython.exceptval(check=True). In this case Cython must always briefly re-acquire the GIL after a function call to check if an exception has been raised. This can commonly happen with a function returning nothing (C void). Simple workarounds are to mark the function as noexcept if you’re certain that exceptions cannot be thrown, or to change the return type to int and just let Cython use the return value as an error flag (by default, -1 triggers the exception check).

Reply:
> Cython definitely has a way of mentioning things in Reverse Polish notation. This is as confusing as the nogil keyword that with gil means gil is on, but nogil means it, maybe, might be off. Same with noexcept.

Yes GIL handling in cython can be easily confusing and I got lost in it multiple times.

> It is not that I have no exceptions possibly raised; but I don't know what noexcept means anymore. 

`noexcept` keyword means that you don't want to propagate the exception from function. It's the same as in Cython 0.29.X when you does not provide `except` keyword.

Consider following function compiled in Cython 0.29.X:

```cython
cdef int sum(int a, int b):
    raise ValueError
```

it is equal to following example (compiled in both Cython 0.29.X and Cython 3)

```cython
cdef int sum(int a, int b) noexcept:
    raise ValueError
```

If exception is raised in function body, it is not propagated, so if you write:

```python
try:
     sum(1, 2)
except ValueError:
     print("Got exception")
```

the `print("Got exception")` line is not executed for both examples.

I am sorry that I cannot explain it better :-(

Reply:
Ah no you are too patient :) No problem at all. Sometimes my brain does not cooperate so it's all on me